src/main/java/uk/co/bconline/ndelius/service/impl/DomainEventServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+package uk.co.bconline.ndelius.service.impl;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.co.bconline.ndelius.model.entity.NDParameterEntity;
+import uk.co.bconline.ndelius.model.entity.ReferenceDataEntity;
+import uk.co.bconline.ndelius.model.notification.DomainEvent;
+import uk.co.bconline.ndelius.model.notification.HmppsDomainEvent;
+import uk.co.bconline.ndelius.model.notification.HmppsDomainEventType;
+import uk.co.bconline.ndelius.repository.db.DomainEventRepository;
+import uk.co.bconline.ndelius.repository.db.NDParameterRepository;
+import uk.co.bconline.ndelius.repository.db.ReferenceDataRepository;
+import uk.co.bconline.ndelius.service.DomainEventService;
+import uk.co.bconline.ndelius.transformer.DomainEventTransformer;
+
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Optional;
+
+@Slf4j
+@Service
+public class DomainEventServiceImpl implements DomainEventService
+{
+    private final ReferenceDataRepository referenceDataRepository;
+
+    private final DomainEventRepository domainEventRepository;
+    private final DomainEventTransformer domainEventTransformer;
+
+    private final NDParameterRepository ndParameterRepository;
+
+    private static final String ENABLE_DOMAIN_EVENTS_PARAM = "ENABLE_DOMAIN_EVENTS";
+    private static final String SELF_PUBLISH_DOMAIN_EVENTS_PARAM = "SELF_PUBLISH_DOMAIN_EVENTS";
+    private static final String DOMAIN_EVENT_TYPE_REF_DATA_CODE_SET = "DOMAIN EVENT TYPE";
+
+    @Autowired
+    public DomainEventServiceImpl(
+            ReferenceDataRepository referenceDataRepository,
+            DomainEventRepository domainEventRepository,
+            DomainEventTransformer domainEventTransformer,
+            NDParameterRepository ndParameterRepository) {
+        this.referenceDataRepository = referenceDataRepository;
+        this.domainEventRepository = domainEventRepository;
+        this.domainEventTransformer = domainEventTransformer;
+        this.ndParameterRepository = ndParameterRepository;
+    }
+
+    @Override
+    public void insertDomainEvent(HmppsDomainEventType eventType, HashMap<String, String> attributes)
+    {
+        if (isDomainEventEnabled())
+        {
+            DomainEvent domainEvent = createDomainEvent(eventType, attributes);
+            if (domainEvent != null)
+            {
+                domainEventRepository.save(domainEventTransformer.map(domainEvent));
+            } else
+            {
+                log.debug("Domain event flags are enabled but no domain event could be created for event type: {}", eventType.getEventType());
+            }
+        }
+    }
```

Comment:
With all the other changes, you should be able to simplify it to something like this:

```suggestion
    public void insertDomainEvent(HmppsDomainEventType eventType, Map<String, String> additionalInformation)
    {
        val type = referenceDataRepository.findByCodeAndReferenceDataMasterCodeSetName(eventType.getEventType(), DOMAIN_EVENT_TYPE_REF_DATA_CODE_SET)
            .orElseThrow(new IllegalStateException("Reference data for domain event type " + eventType.getEventType() + " not found"));
        val message = HmppsDomainEvent.builder()
              .eventType(eventType.getEventType())
              .description(eventType.getEventDescription())
              .occurredAt(ZonedDateTime.now().format(DateTimeFormatter.ISO_ZONED_DATE_TIME))
              .additionalInformation(additionalInformation)
              .version(1)
              .build();
        val attributes = Map.of("eventType", Map.of("Type", "String", "Value", eventType.getEventType()));
        val entity = DomainEventEntity.builder()
              .messageBody(objectMapper.writeValueAsString(message))
              .messageAttributes(objectMapper.writeValueAsString(attributes))
              .domainEventTypeId(type.getId())
              .createdDateTime(LocalDateTime.now())
              .build()
        domainEventRepository.save(entity);
    }
```